raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -341,37 +356,59 @@ private VoteResponseData handleVoteRequest(
             return buildVoteResponse(Errors.INVALID_REQUEST, false);
         }
 
+        // we already confirmed that candidateEpoch >= quorum.epoch()
+        final boolean voteGranted;
         if (candidateEpoch > quorum.epoch()) {
-            // If the epoch is larger than our own, then we need to bump our own epoch,
-            // but we leave the vote undecided until we check the offset below
-            becomeUnattachedFollower(candidateEpoch);
-        }
+            final boolean isLeaderOrcandidate = quorum.isLeader() || quorum.isCandidate();
 
-        if (quorum.isLeader()) {
-            logger.debug("Ignoring vote request {} with epoch {} since we are already the leader in this epoch",
-                    request, candidateEpoch);
-            return buildVoteResponse(Errors.NONE, false);
-        } else if (quorum.isCandidate()) {
-            logger.debug("Ignoring vote request {} with epoch {} since we are already a candidate in this epoch",
-                    request, candidateEpoch);
-            return buildVoteResponse(Errors.NONE, false);
+            OffsetAndEpoch lastEpochEndOffsetAndEpoch = new OffsetAndEpoch(
+                    lastEpochEndOffset, lastEpoch);
+            voteGranted = lastEpochEndOffsetAndEpoch.compareTo(endOffset()) >= 0;
+
+            // even if we rejected the vote, we'd still need to bump our epoch,
+            // however, if the old state is leader or candidate, we should restart election right after the bump
```

Comment:
Can you elaborate on why this makes sense? Why would the fact that we were a candidate or leader before influence this?

Reply:
Hmm... Still not sure I follow why this logic should be different if we were previously a leader or candidate. If I understand correctly, your intuition is that we are more likely to succeed as a candidate because we rejected the vote here (so at least we know our log is ahead of one other member). But that could be true even if we were a follower. 

In general, I think my expectation is that we should set the election timeout in either case if we get a vote request with a larger epoch. The fact that we were a leader previously does not guarantee that we can win the election or that another leader cannot emerge. If we become a candidate too quickly, we may not be giving a chance for that leader to emerge.

Reply:
Yeah I think that's debatable indeed. When received a vote request:

* If the vote is granted, we should be a voted follower with timer reset to election timeout.
* If the voter is not granted since epoch is smaller, then we should not make any changes and not reset timer either.

* The open question is, if the voter is not granted but the epoch is larger, what should we do?

I think if the previous state is leader / candidate, then we should definitely transit to candidate with timer reset to `election`; if the previous state is a follower, however, I think we should either do the same (which would effectively shorten our existing timer from a longer fetch timeout to election timeout), or let that follower stay in unattached follower state while the timer un-touched (i.e. still elapsing on fetch timeout). It's a bit inconsistent if we let the `state` become `follower` while the timer is reset to `election timeout`.

Reply:
@hachikuji I removed the pre-state condition to always transit to unattached-follower while resetting timer to `jitter`.